hal/src/sercom/v2/dma.rs

```diff
@@ -0,0 +1,271 @@
+    C::Word: Beat,
+    D: Capability,
+{
+    type Beat = C::Word;
+
+    #[inline]
+    fn dma_ptr(&mut self) -> *mut Self::Beat {
+        unsafe {
+            #[cfg(feature = "min-samd51g")]
+            {
+                self.sercom().usart_int().data.as_ptr() as *mut _
+            }
+
+            #[cfg(any(feature = "samd11", feature = "samd21"))]
+            {
+                self.sercom().usart().data.as_ptr() as *mut _
+            }
+        }
+    }
+
+    #[inline]
+    fn incrementing(&self) -> bool {
+        false
+    }
+
+    #[inline]
+    fn buffer_len(&self) -> usize {
+        1
+    }
+}
+
+impl<C, D> Uart<C, D>
+where
+    Self: Buffer<Beat = C::Word>,
+    C: uart::ValidConfig,
+    D: Receive,
+{
+    /// Transform an [`Uart`] into a DMA [`Transfer`]) and
+    /// start receiving into the provided buffer.
+    #[inline]
+    pub fn receive_with_dma<Ch, B, W>(
+        self,
+        buf: B,
+        mut channel: Ch,
+        waker: W,
+    ) -> Transfer<Channel<Ch::Id, Busy>, BufferPair<Self, B>, W>
+    where
+        Ch: AnyChannel<Status = Ready>,
+        B: Buffer<Beat = C::Word> + 'static,
+        W: FnOnce(CallbackStatus) + 'static,
+    {
+        channel
+            .as_mut()
+            .enable_interrupts(InterruptFlags::new().with_tcmpl(true));
+
+        #[cfg(feature = "min-samd51g")]
+        let trigger_action = TriggerAction::BURST;
+
+        #[cfg(any(feature = "samd11", feature = "samd21"))]
+        let trigger_action = TriggerAction::BEAT;
+
+        // SAFETY: We use new_unchecked to avoid having to pass a 'static self as the
+        // destination buffer. This is safe as long as we guarantee the source buffer is
+        // static.
+        unsafe { dmac::Transfer::new_unchecked(channel, self, buf, false) }
+            .with_waker(waker)
+            .begin(C::Sercom::DMA_RX_TRIGGER, trigger_action)
+    }
+}
+
+impl<C, D> Uart<C, D>
+where
+    Self: Buffer<Beat = C::Word>,
+    C: uart::ValidConfig,
+    D: Transmit,
+{
+    /// Transform an [`Uart`] into a DMA [`Transfer`]) and
+    /// start sending the provided buffer.
+    #[inline]
+    pub fn send_with_dma<Ch, B, W>(
+        self,
+        buf: B,
+        mut channel: Ch,
+        waker: W,
+    ) -> Transfer<Channel<Ch::Id, Busy>, BufferPair<B, Self>, W>
+    where
+        Ch: AnyChannel<Status = Ready>,
+        B: Buffer<Beat = C::Word> + 'static,
+        W: FnOnce(CallbackStatus) + 'static,
+    {
+        channel
+            .as_mut()
+            .enable_interrupts(InterruptFlags::new().with_tcmpl(true));
+
+        #[cfg(feature = "min-samd51g")]
+        let trigger_action = TriggerAction::BURST;
+
+        #[cfg(any(feature = "samd11", feature = "samd21"))]
+        let trigger_action = TriggerAction::BEAT;
+
+        // SAFETY: We use new_unchecked to avoid having to pass a 'static self as the
+        // destination buffer. This is safe as long as we guarantee the source buffer is
+        // static.
```

Comment:
Shouldn't this say "as long as the *destination* buffer is static"?

Reply:
No, a transmission is memory -> SERCOM. The destination buffer is obviously `'static`, we need to guarantee that the buffer in memory is static. Though you pointing that out made me realize I forgot the static bounds for B and W!

Reply:
Ah yes, but you have the same comment for both the `send_with` and `receive_with` versions, so one of them needs to be changed.

Reply:
Wait, you *do* have the `'static` bounds for `B` and `W`. Am I missing something?

Reply:
Whoops, you're right. Disregard, I was looking at the wrong code section.